src/serialization/serialization.cpp

```diff
@@ -0,0 +1,142 @@
+    header.version = htonl(header.version);
+    header.num_positions = htonl(header.num_positions);
+    header.crc = htonl(header.crc);
+}
+
+/// Converts EulerAngles from host to network byte order.
+void hton(EulerAngles& angles) {
+    angles.roll = htond(angles.roll);
+    angles.ra = htond(angles.ra);
+    angles.de = htond(angles.de);
+}
+
+/// Converts a Vec3 from host to network byte order.
+void hton(Vec3& v) {
+    v.x = htond(v.x);
+    v.y = htond(v.y);
+    v.z = htond(v.z);
+}
+
+/// Converts a DataFileHeader from network to host byte order.
+void ntoh(DataFileHeader& header) {
+    header.version = ntohl(header.version);
+    header.num_positions = ntohl(header.num_positions);
+    header.crc = ntohl(header.crc);
+}
+
+/// Converts EulerAngles from network to host byte order.
+void ntoh(EulerAngles& angles) {
+    angles.roll = ntohd(angles.roll);
+    angles.ra = ntohd(angles.ra);
+    angles.de = ntohd(angles.de);
+}
+
+/// Converts a Vec3 from network to host byte order.
+void ntoh(Vec3& v) {
+    v.x = ntohd(v.x);
+    v.y = ntohd(v.y);
+    v.z = ntohd(v.z);
+}
+
+/// Calculates the CRC32 checksum for a block of memory.
+uint32_t calculateCRC32(const void* data, size_t length) {
+    uint32_t crc = 0xFFFFFFFFU;
+    const uint8_t* bytes = static_cast<const uint8_t*>(data);
+    for (size_t i = 0; i < length; ++i) {
+        crc ^= bytes[i];
+        for (int j = 0; j < 8; ++j) {
+            if (crc & 1)
+                crc = (crc >> 1) ^ 0xEDB88320U;
+            else
+                crc = crc >> 1;
+        }
+    }
+    return crc ^ 0xFFFFFFFFU;
+}
+
+/// Serializes a DataFile object to an output stream.
+void serialize(const DataFile& data, std::ostream& stream) {
+    DataFileHeader header = data.header;
+    hton(header);
+    header.crc = calculateCRC32(&header, sizeof(header) - sizeof(header.crc));
+    header.crc = htonl(header.crc);
+    stream.write(reinterpret_cast<const char*>(&header), sizeof(header));
+
+    EulerAngles relative_attitude = data.relative_attitude;
+    hton(relative_attitude);
+    stream.write(reinterpret_cast<const char*>(&relative_attitude), sizeof(relative_attitude));
+
+    for (auto& r : data.positions) {
+        LocationRecord record = r;
+        hton(record.position);
+        record.timestamp = htonl(record.timestamp);
+        stream.write(reinterpret_cast<const char*>(&record), sizeof(LocationRecord));
+    }
+}
+
+/// Deserializes a DataFile object from an input stream.
+DataFile deserialize(std::istream& stream) {
+    DataFile data;
+    data.header = readHeader(stream);
+
+    stream.read(reinterpret_cast<char*>(&data.relative_attitude), sizeof(EulerAngles));
+    if (stream.gcount() != sizeof(EulerAngles)) {
+        throw std::ios_base::failure("Failed to read relative_attitude");
+    }
+    ntoh(data.relative_attitude);
+
+    data.positions.reserve(data.header.num_positions);
+    for (uint32_t i = 0; i < data.header.num_positions; ++i) {
+        LocationRecord record;
```

Comment:
Then, here, iterate through your array instead of creating a temporary copy, filling the data in each entry. So, roughly:

```
read(reinterpret_cast<char*>(&data.positions[i]), sizeof(LocationRecord));
// error check
ntoh(data.positions[i]);
```

which is simpler now